src/Functions/IFunction.cpp

```diff
@@ -434,8 +432,8 @@ DataTypePtr IFunctionOverloadResolver::getReturnType(const ColumnsWithTypeAndNam
 
         auto type_without_low_cardinality = getReturnTypeWithoutLowCardinality(args_without_low_cardinality);
 
-        if (canBeExecutedOnLowCardinalityDictionary() && has_low_cardinality
-            && num_full_low_cardinality_columns <= 1 && num_full_ordinary_columns == 0
+        if (canBeExecutedOnLowCardinalityDictionary()
+            && num_full_low_cardinality_columns == 1 && num_full_ordinary_columns == 0
```

Comment:
This change means that the resulting type of a function changes when we pass a constant LowCardinality column instead of a full one. Which sounds a bit counter-intuitive...

```
> SELECT concat(toLowCardinality(materialize('x')), '1') as x, toTypeName(x)
x1	LowCardinality(String)
```

```
> SELECT concat(toLowCardinality('x'), '1') as x, toTypeName(x)
x1 String
```

Reply:
The rule for this optimization is simple but I've stuck upon it for a while in recent days: 
if there is a function f(a_1,..,a_n) and all of the a_i are constants except one which is LowCardinality - we can say that the result will be also LowCardinality.

Reply:
Ok, if this is more likely to be confusing in practice, I'll revert it.